pkg/runtime/scheme.go

```diff
@@ -75,111 +130,218 @@ func NewScheme() *Scheme {
 func (s *Scheme) AddKnownTypes(gv unversioned.GroupVersion, types ...Object) {
-	interfaces := make([]interface{}, len(types))
-	for i := range types {
-		interfaces[i] = types[i]
+	if len(gv.Version) == 0 {
+		panic(fmt.Sprintf("version is required on all types: %s %v", gv, types[0]))
+	}
+	for _, obj := range types {
+		t := reflect.TypeOf(obj)
+		if t.Kind() != reflect.Ptr {
+			panic("All types must be pointers to structs.")
+		}
+		t = t.Elem()
+		if t.Kind() != reflect.Struct {
+			panic("All types must be pointers to structs.")
+		}
+
+		gvk := gv.WithKind(t.Name())
+		s.gvkToType[gvk] = t
+		s.typeToGVK[t] = append(s.typeToGVK[t], gvk)
 	}
-	s.raw.AddKnownTypes(gv, interfaces...)
-}
-
-// AddIgnoredConversionType declares a particular conversion that should be ignored - during conversion
-// this method is not invoked.
-func (s *Scheme) AddIgnoredConversionType(from, to interface{}) error {
-	return s.raw.AddIgnoredConversionType(from, to)
 }
 
 // AddKnownTypeWithName is like AddKnownTypes, but it lets you specify what this type should
 // be encoded as. Useful for testing when you don't want to make multiple packages to define
 // your structs.
 func (s *Scheme) AddKnownTypeWithName(gvk unversioned.GroupVersionKind, obj Object) {
-	s.raw.AddKnownTypeWithName(gvk, obj)
+	t := reflect.TypeOf(obj)
+	if len(gvk.Version) == 0 {
+		panic(fmt.Sprintf("version is required on all types: %s %v", gvk, t))
+	}
+	if t.Kind() != reflect.Ptr {
+		panic("All types must be pointers to structs.")
+	}
+	t = t.Elem()
+	if t.Kind() != reflect.Struct {
+		panic("All types must be pointers to structs.")
+	}
+
+	s.gvkToType[gvk] = t
+	s.typeToGVK[t] = append(s.typeToGVK[t], gvk)
 }
 
 // KnownTypes returns the types known for the given version.
-// Return value must be treated as read-only.
 func (s *Scheme) KnownTypes(gv unversioned.GroupVersion) map[string]reflect.Type {
-	return s.raw.KnownTypes(gv)
+	types := make(map[string]reflect.Type)
+	for gvk, t := range s.gvkToType {
+		if gv != gvk.GroupVersion() {
+			continue
+		}
+
+		types[gvk.Kind] = t
+	}
+	return types
 }
 
-// ObjectKind returns the default group,version,kind of the given Object.
+// ObjectKind returns the group,version,kind of the go object,
+// or an error if it's not a pointer or is unregistered.
 func (s *Scheme) ObjectKind(obj Object) (unversioned.GroupVersionKind, error) {
-	return s.raw.ObjectKind(obj)
+	gvks, err := s.ObjectKinds(obj)
+	if err != nil {
+		return unversioned.GroupVersionKind{}, err
+	}
+	return gvks[0], nil
 }
 
-// ObjectKinds returns the all possible group,version,kind of the given Object.
+// ObjectKinds returns all possible group,version,kind of the go object,
+// or an error if it's not a pointer or is unregistered.
 func (s *Scheme) ObjectKinds(obj Object) ([]unversioned.GroupVersionKind, error) {
-	return s.raw.ObjectKinds(obj)
+	v, err := conversion.EnforcePtr(obj)
+	if err != nil {
+		return nil, err
+	}
+	t := v.Type()
+
+	gvks, ok := s.typeToGVK[t]
```

Comment:
We don't allow it, expect it, or do it, but we don't disallow it.

On Fri, Jan 29, 2016 at 6:13 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/runtime/scheme.go
> <https://github.com/kubernetes/kubernetes/pull/20127#discussion_r51328873>
> :
>
> >  func (s *Scheme) ObjectKinds(obj Object) ([]unversioned.GroupVersionKind, error) {
> > -	return s.raw.ObjectKinds(obj)
> > +	v, err := conversion.EnforcePtr(obj)
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	t := v.Type()
> > +
> > +	gvks, ok := s.typeToGVK[t]
>
> Should there be a lock on this map? Do we disallow non-init time
> registrations?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20127/files#r51328873>.
>